File: src/main/java/com/polygloat/service/ApiKeyService.java
package com.polygloat.service;

import com.polygloat.constants.ApiScope;
import com.polygloat.dtos.response.ApiKeyDTO.ApiKeyDTO;
import com.polygloat.exceptions.NotFoundException;
import com.polygloat.model.ApiKey;
import com.polygloat.model.Repository;
import com.polygloat.model.UserAccount;
import com.polygloat.repository.ApiKeyRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.math.BigInteger;
import java.security.SecureRandom;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class ApiKeyService {

    private final ApiKeyRepository apiKeyRepository;
    private final PermissionService permissionService;
    private final SecureRandom random;

    public ApiKeyDTO createApiKey(UserAccount userAccount, Set<ApiScope> scopes, Repository repository) {
        ApiKey apiKey = ApiKey.builder()
                .key(new BigInteger(130, random).toString(32))
                .scopes(scopes)
                .repository(repository)
                .userAccount(userAccount).build();
        apiKeyRepository.save(apiKey);
        return ApiKeyDTO.fromEntity(apiKey);
    }

    public Set<ApiKey> getAllByUser(UserAccount userAccount) {
        return apiKeyRepository.getAllByUserAccountOrderById(userAccount);
    }

    public Set<ApiKey> getAllByRepository(Long repositoryId) {
        return apiKeyRepository.getAllByRepositoryId(repositoryId);
    }

    public Optional<ApiKey> getApiKey(String apiKey) {
        return apiKeyRepository.findByKey(apiKey);
    }

    public Optional<ApiKey> getApiKey(Long id) {
        return apiKeyRepository.findById(id);
    }

    public void deleteApiKey(ApiKey apiKey) {
        apiKeyRepository.delete(apiKey);
    }

    public Set<ApiScope> getAvailableScopes(UserAccount userAccount, Repository repository) {
        return Arrays.stream(
                permissionService.getRepositoryPermission(repository.getId(), userAccount).orElseThrow(NotFoundException::new).getType().getAvailableScopes()
        ).collect(Collectors.toSet());
    }

    public void editApiKey(ApiKey apiKey) {
        apiKeyRepository.save(apiKey);
    }

    public void deleteAllByRepository(Long repositoryId) {
        apiKeyRepository.deleteAllByRepositoryId(repositoryId);
    }
}

File: backend/data/src/main/kotlin/io/tolgee/service/ApiKeyService.kt
package io.tolgee.service

import io.tolgee.exceptions.NotFoundException
import io.tolgee.model.ApiKey
import io.tolgee.model.Project
import io.tolgee.model.UserAccount
import io.tolgee.model.enums.ApiScope
import io.tolgee.repository.ApiKeyRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import java.math.BigInteger
import java.security.SecureRandom
import java.util.*

@Service
class ApiKeyService @Autowired constructor(
  private val apiKeyRepository: ApiKeyRepository,
  private val random: SecureRandom
) {

  @set:Autowired
  lateinit var permissionService: PermissionService

  fun create(userAccount: UserAccount, scopes: Set<ApiScope>, project: Project): ApiKey {
    val apiKey = ApiKey(
      key = BigInteger(130, random).toString(32),
      project = project,
      userAccount = userAccount,
      scopesEnum = scopes
    )
    return apiKeyRepository.save(apiKey)
  }

  fun getAllByUser(userAccountId: Long): Set<ApiKey> {
    return apiKeyRepository.getAllByUserAccountIdOrderById(userAccountId)
  }

  fun getAllByUser(userAccountId: Long, filterProjectId: Long?, pageable: Pageable): Page<ApiKey> {
    return apiKeyRepository.getAllByUserAccount(userAccountId, filterProjectId, pageable)
  }

  fun getAllByProject(projectId: Long): Set<ApiKey> {
    return apiKeyRepository.getAllByProjectId(projectId)
  }

  fun getAllByProject(projectId: Long, pageable: Pageable): Page<ApiKey> {
    return apiKeyRepository.getAllByProjectId(projectId, pageable)
  }

  fun getApiKey(apiKey: String): Optional<ApiKey> {
    return apiKeyRepository.findByKey(apiKey)
  }

  fun getApiKey(id: Long): Optional<ApiKey> {
    return apiKeyRepository.findById(id)
  }

  fun deleteApiKey(apiKey: ApiKey) {
    apiKeyRepository.delete(apiKey)
  }

  fun getAvailableScopes(userAccountId: Long, project: Project): Set<ApiScope> {
    return permissionService.getProjectPermissionType(project.id, userAccountId)?.availableScopes?.toSet()
      ?: throw NotFoundException()
  }

  fun editApiKey(apiKey: ApiKey): ApiKey {
    return apiKeyRepository.save(apiKey)
  }

  fun deleteAllByProject(projectId: Long) {
    apiKeyRepository.deleteAllByProjectId(projectId)
  }

  fun saveAll(entities: Iterable<ApiKey>) {
    this.apiKeyRepository.saveAll(entities)
  }

  fun save(entity: ApiKey) {
    this.apiKeyRepository.save(entity)
  }
}
